drivers/led/lp50xx.c

```diff
@@ -0,0 +1,408 @@
+		return LP5036_LED0_BRIGHTNESS;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int lp50xx_get_color_base_addr(enum lp50xx_chips id)
+{
+	switch (id) {
+	case LP5009:
+	case LP5012:
+		return LP5012_OUT0_COLOR;
+	case LP5018:
+	case LP5024:
+		return LP5024_OUT0_COLOR;
+	case LP5030:
+	case LP5036:
+		return LP5036_OUT0_COLOR;
+	default:
+		return -EINVAL;
+	}
+}
+
+static const struct led_info *lp50xx_led_to_info(
+			const struct lp50xx_config *config, uint32_t led)
+{
+	const struct led_info *led_info = NULL;
+
+	for (uint8_t i = 0; i < config->num_leds; i++) {
+		if (config->leds_info[i].index == led) {
+			led_info = &config->leds_info[i];
+		}
+	}
+	return led_info;
+}
+
+static int lp50xx_get_info(const struct device *dev, uint32_t led,
+			   const struct led_info **info)
+{
+	const struct lp50xx_config *config = dev->config;
+	const struct led_info *led_info = lp50xx_led_to_info(config, led);
+
+	if (!led_info) {
+		return -EINVAL;
+	}
+
+	*info = led_info;
+
+	return 0;
+}
+
+static int lp50xx_set_brightness(const struct device *dev,
+				 uint32_t led, uint8_t value)
+{
+	const struct lp50xx_config *config = dev->config;
+	const struct led_info *led_info = lp50xx_led_to_info(config, led);
+	uint8_t buf[2];
+
+	if (!led_info) {
+		return -ENODEV;
+	}
+
+	if (!IN_RANGE(value, LP50XX_MIN_BRIGHTNESS, LP50XX_MAX_BRIGHTNESS)) {
+		LOG_ERR("%s: brightness value out of bounds: "
+			"val=%d, min=%d, max=%d",
+			dev->name,
+			value,
+			LP50XX_MIN_BRIGHTNESS,
+			LP50XX_MAX_BRIGHTNESS);
+		return -EINVAL;
+	}
+
+	buf[0] = lp50xx_get_brightness_base_addr(config->chip) + led_info->index;
+	buf[1] = (value * 0xff) / 100;
+
+	return i2c_write_dt(&config->bus, buf, sizeof(buf));
+}
+
+static int lp50xx_on(const struct device *dev, uint32_t led)
+{
+	return lp50xx_set_brightness(dev, led, 100);
+}
+
+static int lp50xx_off(const struct device *dev, uint32_t led)
+{
+	return lp50xx_set_brightness(dev, led, 0);
+}
+
+static int lp50xx_set_color(const struct device *dev, uint32_t led,
+			    uint8_t num_colors, const uint8_t *color)
+{
+	const struct lp50xx_config *config = dev->config;
+	const struct led_info *led_info = lp50xx_led_to_info(config, led);
+	uint8_t buf[4];
+
+	if (!led_info) {
+		return -ENODEV;
+	}
+
+	if (num_colors != led_info->num_colors) {
+		LOG_ERR("%s: invalid number of colors: got=%d, expected=%d",
+			dev->name,
+			num_colors,
+			led_info->num_colors);
+		return -EINVAL;
+	}
+
+	buf[0] = lp50xx_get_color_base_addr(config->chip) + 3 * led_info->index;
+	buf[1] = color[0];
+	buf[2] = color[1];
+	buf[3] = color[2];
+
+	return i2c_write_dt(&config->bus, buf, sizeof(buf));
+}
+
+static int lp50xx_write_channels(const struct device *dev,
+				 uint32_t start_channel,
+				 uint32_t num_channels, const uint8_t *buf)
+{
+	const struct lp50xx_config *config = dev->config;
+	struct lp50xx_data *data = dev->data;
+	uint8_t base_channel, end_channel, max_channels;
+
+	base_channel = lp50xx_get_brightness_base_addr(config->chip);
```

Comment:
NAK. Before this patch the base channel register of the `write_channels` API is "bank brightness". Now it becomes "LED0 brightness". If there are apps using this driver out there, then this change breaks them silently (without even giving an explanation or a warning it in the commit message).

So please remove this change from this patch and even better from this PR. Let's make this PR just be about adding support for more controller models. If needed, you'll can change `lp50xx_write_channels` with another PR.

Reply:
OK, let's not break things

Reply:
After thought, that is why I find the `write_channels()` API difficult to work with for this controller. Since it is giving raw register access, it is necessary to expose it in some way. Since the base addresses are not the same for all chips, things are gonna break or we need to find a way to expose a common macro to use at runtime and rely on the chip id (and ternary operators).

Even if I don't like it, I think I will expose these macros in `include/zephyr/drivers/led/lp50xx.h` for now just to merge this PR and move on to simplify it in the future.

Reply:
The base address is different but it is handled by the driver. For the application, channels 0, 1, 2 and 3 are "bank brightness", "bank color A", "bank color B" and "bank color C" registers. Is is true for all models, isn't it ? The driver translates the channel index into a register address. At least it is the current implementation. Can't we do the same with more controller models ?

After the LED brightness and LED color channels indexes can be computed using the maximum number of LEDs (which depends on the chip type).

Reply:
> After the LED brightness and LED color channels indexes can be computed using the maximum number of LEDs (which depends on the chip type).

Yes and no, but I am working on it using an extra offset information.
For example, LP5018 and LP5024 share the same register addresses but LP5018 has a maximum of 6 leds and LP5024 has 8. So we can't directly use the chip type and the maximum number of leds to calculate an offset.